File: app/src/main/java/com/bonstead/pitdroid/MainActivity.java
package com.bonstead.pitdroid;

import java.lang.ref.WeakReference;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

import android.annotation.TargetApi;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.View;
import android.view.WindowManager;
import android.widget.Toast;

public class MainActivity extends Activity implements OnSharedPreferenceChangeListener
{
	static final String TAG = "MainActivity";

	private final ScheduledExecutorService mScheduler = Executors.newScheduledThreadPool(1);
	private ScheduledFuture<?> mUpdateTimer = null;
	private HeaterMeter mHeaterMeter = null;
	private boolean mAllowServiceShutdown = false;

	private final Runnable mUpdate = new Runnable()
	{
		public void run()
		{
			Object data = mHeaterMeter.updateThread();
			mHandler.sendMessage(mHandler.obtainMessage(0, data));
		}
	};

	@Override
	public void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "onCreate");
		}

		// If we don't have a fragment (being opened, not recreated), open the gauge fragment by default
		if (getFragmentManager().findFragmentById(android.R.id.content) == null)
		{
			// Display the fragment as the main content.
			getFragmentManager().beginTransaction()
					.add(android.R.id.content, new GaugeFragment())
					.commit();
		}

		mHeaterMeter = ((PitDroidApplication) this.getApplication()).mHeaterMeter;

		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getBaseContext());
		mHeaterMeter.initPreferences(prefs);
		prefs.registerOnSharedPreferenceChangeListener(this);

		// Uncomment to use saved sample data instead of live, for testing purposes
		//mHeaterMeter.setHistory(new InputStreamReader(getResources().openRawResource(R.raw.sample_data)));

		updateScreenOn();
		updateAlarmService();

		// Sent when the close button is pressed on the alarm service status message
		if (getIntent().hasExtra("close"))
		{
			showCloseMessage();
		}
	}

	@Override
	protected void onPause()
	{
		super.onPause();

		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "onPause");
		}

		if (mUpdateTimer != null)
		{
			if (BuildConfig.DEBUG)
			{
				Log.v(TAG, "Canceling update timer");
			}

			mUpdateTimer.cancel(false);
			mUpdateTimer = null;
		}
	}

	@Override
	protected void onPostResume()
	{
		super.onPostResume();

		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "onPostResume");
		}

		if (mUpdateTimer == null)
		{
			if (BuildConfig.DEBUG)
			{
				Log.v(TAG, "Starting update timer");
			}

			mUpdateTimer = mScheduler.scheduleAtFixedRate(mUpdate, 0, HeaterMeter.kMinSampleTime, TimeUnit.MILLISECONDS);
		}
		else
		{
			if (BuildConfig.DEBUG)
			{
				Log.v(TAG, "Update timer already set, skipping");
			}
		}
	}

	@Override
	protected void onDestroy()
	{
		super.onDestroy();

		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "onDestroy");
		}

		SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getBaseContext());
		prefs.unregisterOnSharedPreferenceChangeListener(this);

		if (mAllowServiceShutdown)
		{
			stopAlarmService();
		}
	}

	public void onSettingsClick(View v)
	{
		getFragmentManager().beginTransaction()
			.replace(android.R.id.content, new SettingsFragment())
			.addToBackStack(null)
			.commit();
	}

	public void onGraphClick(View v)
	{
		getFragmentManager().beginTransaction()
			.replace(android.R.id.content, new GraphFragment())
			.addToBackStack(null)
			.commit();
	}

	public void onDashClick(View v)
	{
		getFragmentManager().beginTransaction()
			.replace(android.R.id.content, new DashFragment())
			.addToBackStack(null)
			.commit();
	}

	static class IncomingHandler extends Handler
	{
		private final WeakReference<MainActivity> mActivity;

		IncomingHandler(MainActivity activity)
		{
			mActivity = new WeakReference<MainActivity>(activity);
		}

		@Override
		public void handleMessage(Message msg)
		{
			MainActivity activity = mActivity.get();
			if (activity != null)
			{
				activity.mHeaterMeter.updateMain(msg.obj);

				if (activity.mHeaterMeter.mLastStatusMessage != null)
				{
					Context context = activity.getApplicationContext();
					CharSequence text = activity.mHeaterMeter.mLastStatusMessage;
					int duration = Toast.LENGTH_SHORT;

					Toast toast = Toast.makeText(context, text, duration);
					toast.show();

					activity.mHeaterMeter.mLastStatusMessage = null;
				}
			}
		}
	}

	private final IncomingHandler mHandler = new IncomingHandler(this);

	public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key)
	{
		// If any preferences change, have the HeaterMeter re-read them all
		mHeaterMeter.initPreferences(sharedPreferences);
		updateScreenOn();
	}

	private void updateScreenOn()
	{
		if (mHeaterMeter.mKeepScreenOn)
		{
			getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
		}
		else
		{
			getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
		}
	}

	public void updateAlarmService()
	{
		if (mHeaterMeter.hasAlarms())
		{
			stopAlarmService();
			startAlarmService();
		}
		else
		{
			stopAlarmService();
		}
	}

	@TargetApi(Build.VERSION_CODES.O)
	private void startAlarmService()
	{
		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "Start alarm service");
		}

		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
		{
			startForegroundService(new Intent(this, AlarmService.class));
		}
		else
		{
			startService(new Intent(this, AlarmService.class));
		}
		startService(new Intent(this, AlarmService.class));
	}

	private void stopAlarmService()
	{
		if (BuildConfig.DEBUG)
		{
			Log.v(TAG, "Stop alarm service");
		}

		stopService(new Intent(this, AlarmService.class));
	}

	private void showCloseMessage()
	{
		AlertDialog.Builder builder = new AlertDialog.Builder(this);

		builder.setTitle("Confirm");
		builder.setMessage("You have alarms set, are you sure you want to exit?");

		builder.setPositiveButton("Yes", new DialogInterface.OnClickListener()
		{
			public void onClick(DialogInterface dialog, int which)
			{
				mAllowServiceShutdown = true;
				dialog.dismiss();
				finish();
			}
		});

		builder.setNegativeButton("No", null);

		AlertDialog alert = builder.create();
		alert.show();
	}
}

File: app/src/main/java/com/bonstead/pitdroid/MainActivity.kt
package com.bonstead.pitdroid

import java.lang.ref.WeakReference
import android.os.Bundle
import com.google.android.material.bottomnavigation.BottomNavigationView
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import android.content.SharedPreferences
import androidx.preference.PreferenceManager
import kotlinx.android.synthetic.main.activity_main.*
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import android.os.Handler
import android.view.WindowManager
import android.widget.Toast
import android.os.Message
import android.annotation.TargetApi
import android.os.Build
import android.content.Intent
import android.app.AlertDialog

class MainActivity : AppCompatActivity(), SharedPreferences.OnSharedPreferenceChangeListener {
    private val mScheduler = Executors.newScheduledThreadPool(1)
    private var mUpdateTimer: ScheduledFuture<*>? = null
    private var mAllowServiceShutdown = false
    private val mHandler = IncomingHandler(this)

    private val mUpdate = Runnable {
        val data = HeaterMeter.updateThread()
        mHandler.sendMessage(mHandler.obtainMessage(0, data))
    }

    private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
        when (item.itemId) {
            R.id.navigation_dash -> {
                openFragment(DashFragment())
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_graph -> {
                openFragment(GraphFragment())
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_gauge -> {
                openFragment(GaugeFragment())
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_settings -> {
                openFragment(SettingsFragment())
                return@OnNavigationItemSelectedListener true
            }
        }
        false
    }

    private fun openFragment(fragment: Fragment) {
        val transaction = supportFragmentManager.beginTransaction()
        transaction.replace(R.id.fragment, fragment)
        transaction.addToBackStack(null)
        transaction.commit()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // If we don't have a fragment (being opened, not recreated), open the gauge fragment by default
        if (supportFragmentManager.findFragmentById(android.R.id.content) == null) {
            // Display the fragment as the main content.
            openFragment(GaugeFragment())
        }

        val prefs = PreferenceManager.getDefaultSharedPreferences(baseContext)
        prefs.registerOnSharedPreferenceChangeListener(this);

        HeaterMeter.initPreferences(prefs)

        // Uncomment to use saved sample data instead of live, for testing purposes
        //mHeaterMeter.setHistory(new InputStreamReader(getResources().openRawResource(R.raw.sample_data)));

        updateScreenOn()
        updateAlarmService()

        // Sent when the close button is pressed on the alarm service status message
        if (intent.hasExtra("close")) {
            showCloseMessage()
        }

        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)
    }

    override fun onPause() {
        super.onPause()

        if (mUpdateTimer != null) {
            mUpdateTimer!!.cancel(false)
            mUpdateTimer = null
        }
    }

    override fun onPostResume() {
        super.onPostResume()

        if (mUpdateTimer == null) {
            mUpdateTimer = mScheduler.scheduleAtFixedRate(mUpdate, 0, HeaterMeter.kMinSampleTime, TimeUnit.MILLISECONDS)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        val prefs = PreferenceManager.getDefaultSharedPreferences(baseContext)
        prefs.unregisterOnSharedPreferenceChangeListener(this)

        if (mAllowServiceShutdown) {
            stopAlarmService()
        }
    }

    internal class IncomingHandler(activity: MainActivity) : Handler() {
        private val mActivity: WeakReference<MainActivity> = WeakReference(activity)

        override fun handleMessage(msg: Message) {
            val activity = mActivity.get()
            if (activity != null) {
                HeaterMeter.updateMain(msg.obj)

                if (HeaterMeter.mLastStatusMessage != null) {
                    val context = activity.applicationContext
                    val text = HeaterMeter.mLastStatusMessage
                    val duration = Toast.LENGTH_SHORT

                    val toast = Toast.makeText(context, text, duration)
                    toast.show()

                    HeaterMeter.mLastStatusMessage = null
                }
            }
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        // If any preferences change, have the HeaterMeter re-read them all
        HeaterMeter.initPreferences(sharedPreferences)
        updateScreenOn()
    }

    private fun updateScreenOn() {
        if (HeaterMeter.mKeepScreenOn) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    fun updateAlarmService() {
        if (HeaterMeter.hasAlarms()) {
            stopAlarmService()
            startAlarmService()
        } else {
            stopAlarmService()
        }
    }

    @TargetApi(Build.VERSION_CODES.O)
    private fun startAlarmService() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(Intent(this, AlarmService::class.java))
        } else {
            startService(Intent(this, AlarmService::class.java))
        }
        startService(Intent(this, AlarmService::class.java))
    }

    private fun stopAlarmService() {
        stopService(Intent(this, AlarmService::class.java))
    }

    private fun showCloseMessage() {
        val builder = AlertDialog.Builder(this)

        builder.setTitle("Confirm")
        builder.setMessage("You have alarms set, are you sure you want to exit?")

        builder.setPositiveButton("Yes") { dialog, _ ->
            mAllowServiceShutdown = true
            dialog.dismiss()
            finish()
        }

        builder.setNegativeButton("No", null)

        val alert = builder.create()
        alert.show()
    }

    companion object {
        internal val TAG = "MainActivity"
    }
}
